tests/database/odbc-w-enclave/odbc_proxy.c

```diff
@@ -0,0 +1,388 @@
+// CMK and CEK VALUE are used
+#define CMK_NAME "SGXTESTCMK"
+#define CMK_VALUE "abcdefghijklmnopkrstuvwxyz012345"
+#define CEK_NAME "SGXTESTCEK"
+#define CEK_VALUE "012345abcdefghijklmnopkrstuvwxyz"
+#define TABLE_NAME "SGXTESTTABLE"
+#define MAX_KEY_LEN 32       // Max Key Length for both CMK and CEK
+#define DEFAULT_VCHAR_LEN 32 // Default length of all varchar fields
+#define DLLPATH "./cksp.so"
+#define KSPNAME L"AZURE_KEY_VAULT"
+#define SQL_BUFFER_LEN 1024
+
+#define OK_CHECK(EXPRESSION)         \
+    do                               \
+    {                                \
+        int _result_ = (EXPRESSION); \
+        if (_result_ != SUCCESS)     \
+            goto done;               \
+    } while (0)
+
+static void* hProvLib = NULL;
+static SQLHENV env = NULL;
+static SQLHDBC dbc = NULL;
+
+static size_t wchar_str_len(const wchar_t* ws)
+{
+    size_t i = 0;
+    for (const int16_t* p = ws; *p != 0; p++, i++)
+        ;
+    return i;
+}
+
+static int wchar_str_cmp(const wchar_t* p1, const wchar_t* p2)
+{
+    return memcmp(p1, p2, (wchar_str_len(p1) + 1) * sizeof(int16_t));
+}
+
+static void print_sql_info(SQLRETURN rc, SQLHANDLE h, SQLSMALLINT ht)
+{
+    if (!h)
+        return;
+
+    SQLSMALLINT i = 0;
+    SQLSMALLINT outlen = 0;
+    char errmsg[1024];
+    while ((rc = SQLGetDiagField(
+                ht,
+                h,
+                ++i,
+                SQL_DIAG_MESSAGE_TEXT,
+                errmsg,
+                sizeof(errmsg),
+                &outlen)) == SQL_SUCCESS ||
+           rc == SQL_SUCCESS_WITH_INFO)
+    {
+        fprintf(stderr, "Return Code#%d: %s\n", i, errmsg);
+    }
+}
+
+static int checkRC(SQLRETURN rc, char* msg, SQLHANDLE h, SQLSMALLINT ht)
+{
+    if (rc == SQL_SUCCESS_WITH_INFO)
+    {
+        print_sql_info(rc, h, ht);
+    }
+    else if (rc == SQL_ERROR || rc != SQL_SUCCESS)
+    {
+        fprintf(stderr, "Error occurred upon %s\n", msg);
+        print_sql_info(rc, h, ht);
+        return FAILURE;
+    }
+    return SUCCESS;
+}
+
+static int clean_up(SQLHSTMT stmt)
+{
+    int result = FAILURE;
+    SQLRETURN rc;
+    printf("Drop table, CMK and CEK.\n");
+    rc = SQLExecDirect(stmt, "DROP TABLE IF EXISTS " TABLE_NAME, SQL_NTS);
+    OK_CHECK(checkRC(rc, "DROP TABLE.", stmt, SQL_HANDLE_STMT));
+    // do not explicitly check key drops, because if the does not exist, will
+    // return error code
+    SQLExecDirect(stmt, "DROP COLUMN ENCRYPTION KEY " CEK_NAME, SQL_NTS);
+    SQLExecDirect(stmt, "DROP COLUMN MASTER KEY " CMK_NAME, SQL_NTS);
+
+    result = SUCCESS;
+done:
+    return result;
+}
+
+static int create_CMK(SQLHSTMT stmt)
+{
+    int result = FAILURE;
+    SQLRETURN rc;
+    char* stmt_str =
+        "CREATE COLUMN MASTER KEY " CMK_NAME " WITH"
+        "(KEY_STORE_PROVIDER_NAME = N'AZURE_KEY_VAULT',"
+        "KEY_PATH = "
+        "N'https://acckeyvault.vault.azure.net/keys/TestCMK1/"
+        "2a89d8e0cb834009827d588a32709e39',"
```

Comment:
Please document where all those magic strings come from. This has to be reproducible with a custom SQL server / AKV instance.

Reply:
Since we are using custom key store provider, this AKV url is just a place holder. But it is required by SQL Server at least for now. The key value will not be actually accessed by the Database Server. @jxyang any other comments?

Reply:
It looks very real to me. Can you not replace that with a dummy url?

Reply:
Theoretically yes, but that requires us to reverse engineer the enclave signature hashing algorithm(based on the given dummy url).

Reply:
I don't understand, how is hashing related to the URL? What happens if you replace the URL?

Reply:
Here is the [CREATE COLUMN MASTER KEY (Transact-SQL)](https://docs.microsoft.com/en-us/sql/t-sql/statements/create-column-master-key-transact-sql?view=sql-server-ver15), to generate a column master key, a `[,ENCLAVE_COMPUTATIONS (SIGNATURE = signature)]` is also required. Usually that signature hash is generated by SQL Server client like(SSMS, Data Studio). The signature is used to verify the authentic of the Key Path URL(dummy URL in our case). You have to manually compute the signature hash in your application if you want to generate pragmatically.

Reply:
OK, could you add some comment to the code that explains this?

Reply:
Again this is a hack to work around the SQL AE's failure to recognize custom key vaults. The key path has to be a real akv path (even though it's a placeholder only and never been accessed for key operations). The signature has to be a hash value computed from the key path (and other settings). Please explain this in the comments. @praenubilus